lib/Reader/readerir.cpp

```diff
@@ -2952,40 +2952,64 @@ IRNode *GenIR::binaryOp(ReaderBaseNS::BinaryOpcode Opcode, IRNode *Arg1,
   return Result;
 }
 
-Type *GenIR::binaryOpType(Type *Type1, Type *Type2) {
-  if (Type1->isPointerTy()) {
-    if (Type2->isPointerTy()) {
-      return Type::getIntNTy(*this->JitContext->LLVMContext,
-                             TargetPointerSizeInBits);
+Type *GenIR::binaryOpType(ReaderBaseNS::BinaryOpcode Opcode, Type *Type1,
+                          Type *Type2) {
+  // Roughly follows ECMA-355, Table III.2.
+  const bool Type1IsInt = Type1->isIntegerTy();
+  const bool Type2IsInt = Type2->isIntegerTy();
+
+  // If both types are integers, the result is the larger integer type.
+  if (Type1IsInt && Type2IsInt) {
+    uint32_t Size1 = Type1->getPrimitiveSizeInBits();
+    uint32_t Size2 = Type2->getPrimitiveSizeInBits();
+    return (Size1 >= Size2 ? Type1 : Type2);
+  }
+
+  // If both types are floats, the result is the larger float type.
+  if (Type1->isFloatingPointTy() && Type2->isFloatingPointTy()) {
+    uint32_t Size1 = Type1->getPrimitiveSizeInBits();
+    uint32_t Size2 = Type2->getPrimitiveSizeInBits();
+    return (Size1 >= Size2 ? Type1 : Type2);
+  }
+
+  // If we see a mixture of int and unmanaged pointer, the result
+  // is generally a native int, with a few special cases where we
+  // preserve pointer-ness.
+  const bool Type1IsUnmanagedPointer = isUnmanagedPointerType(Type1);
+  const bool Type2IsUnmanagedPointer = isUnmanagedPointerType(Type2);
+
+  const bool IsStrictlyAdd = (Opcode == ReaderBaseNS::Add);
+  const bool IsAdd = IsStrictlyAdd || (Opcode == ReaderBaseNS::AddOvf) ||
+                     (Opcode == ReaderBaseNS::AddOvfUn);
+  const bool IsStrictlySub = (Opcode == ReaderBaseNS::Sub);
+  const bool IsSub = IsStrictlySub || (Opcode == ReaderBaseNS::SubOvf) ||
+                     (Opcode == ReaderBaseNS::SubOvfUn);
+  const bool IsStrictlyAddOrSub = IsStrictlyAdd || IsStrictlySub;
+  const bool IsAddOrSub = IsAdd || IsSub;
+
+  if (Type1IsUnmanagedPointer || Type2IsUnmanagedPointer) {
+    // ptr +/- int = ptr
+    if (IsAddOrSub && Type1IsUnmanagedPointer && Type2IsInt) {
+      return Type1;
     }
-    ASSERTNR(!Type2->isFloatingPointTy());
-    return Type1;
-  } else if (Type2->isPointerTy()) {
-    ASSERTNR(!Type1->isFloatingPointTy());
-    return Type2;
-  }
-
-  if (Type1 == Type2) {
-    return Type1;
-  }
-
-  uint32_t Size1 = Type1->getPrimitiveSizeInBits();
-  uint32_t Size2 = Type2->getPrimitiveSizeInBits();
-
-  if (Type1->isFloatingPointTy()) {
-    if (Type2->isFloatingPointTy()) {
-      return Size1 > Size2 ? Type1 : Type2;
+    // int + ptr = ptr
+    if (IsAdd && Type1IsInt && Type2IsUnmanagedPointer) {
+      return Type2;
     }
-    return Type1;
-  } else if (Type2->isFloatingPointTy()) {
-    return Type2;
+    // Otherwise type result as native int.
```

Comment:
On a 32-bit platform, if we see unmanaged pointer and int64, this would truncate the int64.  That seems a little odd, but since Table III.2 indicates that (`native int`, `int64`) and (`int32`, `int64`) are not valid combinations, I think it's ok (in that situation, the unmanaged pointer presumably should have stack type `int32` or `native int`, and the i64 should have stack type `int64).
I think that rejecting those cases would be ok too; whichever way you prefer.